sbi/inference/posteriors/vi_posterior.py

```diff
@@ -1,69 +1,481 @@
         """
-        super().__init__(potential_fn, theta_transform, device)
+        super().__init__(potential_fn, theta_transform, device, x_shape=x_shape)
 
+        if device is None:
+            device = potential_fn.device
+        else:
+            # Especially the prior may be on another device -> move it...
+            self.potential_fn.device = device
+            move_all_tensor_to_device(self.potential_fn, device)
         self._device = device
+        self._prior = self.potential_fn.prior
+        self._optimizer = None
+
+        # In contrast to MCMC we want to project into constrained space!
+        if theta_transform is None:
+            self.link_transform = mcmc_transform(self._prior).inv
+        else:
+            self.link_transform = theta_transform.inv
+
+        # This will set the variational distribution and VI method
+        self.set_q(q, q_kwargs)
+        self.set_vi_method(vi_method)
 
         self._purpose = (
             "It provides Variational inference to .sample() from the posterior and "
             "can evaluate the _normalized_ posterior density with .log_prob()."
         )
 
+    @property
+    def q(self) -> Distribution:
+        """Returns the variational posterior."""
+        return self._q
+
+    @q.setter
+    def q(
+        self,
+        q: Union[str, Distribution, NeuralPosterior],
+    ) -> None:
+        """Sets the variational distribution. If the distribution does not admit access
+        through "parameters" and "modules" function, please use 'set_q' if you want to
+        explicitly specify the parameters and modules.
+
+
+        Args:
+            q: Variational distribution
+
+
+        """
+        self.set_q(q)
+
+    def set_q(
+        self,
+        q: Union[str, Distribution, NeuralPosterior],
+        q_kwargs: Dict = {},
+    ) -> None:
+        """Defines the variational family. You can specify over which parameters/modules
+        we optimize. This is required for custom distributions which e.g. do not inherit
+        nn.Modules or has the function "parameters" or "modules" to
+        give direct access to trainable parameters.
+
+
+
+        Args:
+            q: Variational distribution.
+            parameters: List of parameters associated with the distribution object.
+            modules: List of modules associated with the distribution object.
+
+
+        """
+        self._q_arg = q
+        self._q_kwargs = q_kwargs
+        if isinstance(q, Distribution):
+            q = adapt_variational_distribution(
+                q, q_kwargs, self._prior, self.link_transform
+            )
+            make_object_deepcopy_compatible(q)
+            self._trained_on = None
+        elif isinstance(q, str):
+            q = get_flow_builder(
+                q,
+                self._prior.event_shape,
+                self.link_transform,
+                device=self._device,
+                **q_kwargs,
```

Comment:
Mhh, right. I just also added the possibility to pass "parameters" or "modules" within there if a user defined distribution object is given. Should I add just these as optional arguments to __init__ then?

Reply:
i think so, yes

Reply:
Maybe we also just alow to pass [str, VIPosterior, Callable], whereas the Callable constructs a distribution. This can then also be used to define other distributions. But would be more consistent with 'inference'.